superbuild/CMakeLists.txt

```diff
@@ -32,41 +32,91 @@ check_cache_var_declaration(L0_TAG "Level Zero url")
 option(CCACHE "Enable ccache to speed up repetitive builds" OFF)
 option(FULL_STAGE2_LLVM "Enable full llvm stage2 build" OFF)
 option(LTO "Enable LTO" OFF)
 option(PGO "Enale PGO (implicitly enable LTO)" OFF)
+option(BUILD_STAGE1_TOOLCHAIN_ONLY "Enable build of stage1 toolchain only." OFF)
 option(BUILD_STAGE2_TOOLCHAIN_ONLY "Enable build of stage2 toolchain only, i.e., skip ispc building and stage3." OFF)
 option(BUILD_SPIRV_TRANSLATOR_ONLY "Enable only build of SPIRV-LLVM-Translator" OFF)
 option(BUILD_VC_INTRINSICS_ONLY "Enable only build of vc-intrinsics" OFF)
 option(BUILD_L0_LOADER_ONLY "Enable only build of level-zero loader" OFF)
-option(XE_DEPS "Enable build of XE dependencies (L0 loader, vc-intrinsics, SPIRV-LLVM-Translator" ON)
+option(INSTALL_ISPC "Enable installation of ISPC to CMAKE_INSTALL_PREFIX" OFF)
+option(INSTALL_WITH_XE_DEPS "Enable installation ISPC or stage2 toolchain with XE_DEPS to CMAKE_INSTALL_PREFIX" OFF)
+option(XE_DEPS "Enable build of all XE dependencies (L0 loader, vc-intrinsics, SPIRV-LLVM-Translator" ON)
+option(MACOS_UNIVERSAL_BIN "Enable build of universal binaries for macOS" ON)
 
 # Force LTO when PGO enabled to the sake of simplicity and less number of
 # possible build variants.
 if (PGO)
     set(LTO ON)
 endif()
 
+set(SKIP_STAGE1 OFF)
+if (NOT PREBUILT_STAGE1 STREQUAL "NO" OR NOT PREBUILT_STAGE1_PATH STREQUAL "NO")
+    set(SKIP_STAGE1 ON)
+    set(FULL_STAGE2_LLVM ON)
+endif()
+
+if (NOT PREBUILT_STAGE1 STREQUAL "NO" AND NOT PREBUILT_STAGE1_PATH STREQUAL "NO")
+    message(FATAL_ERROR "Provide only one source of stage1 toolchain.")
+endif()
+
+set(SKIP_STAGE2 OFF)
 if (NOT PREBUILT_STAGE2 STREQUAL "NO" OR NOT PREBUILT_STAGE2_PATH STREQUAL "NO")
-    set(SKIP_STAGE2_DEPS ON)
+    set(SKIP_STAGE2 ON)
 endif()
 
 if (NOT PREBUILT_STAGE2 STREQUAL "NO" AND NOT PREBUILT_STAGE2_PATH STREQUAL "NO")
     message(FATAL_ERROR "Provide only one source of stage2 toolchain.")
 endif()
 
 set(STAGE2_PATH "${CMAKE_BINARY_DIR}/stage2")
+if (NOT PREBUILT_STAGE1_PATH STREQUAL "NO")
+    set(STAGE2_PATH "${PREBUILT_STAGE1_PATH}")
+endif()
 if (NOT PREBUILT_STAGE2_PATH STREQUAL "NO")
     set(STAGE2_PATH "${PREBUILT_STAGE2_PATH}")
 endif()
+set(STAGE3_PATH "${CMAKE_BINARY_DIR}/stage3")
+
+if (NOT PREBUILT_STAGE1_PATH STREQUAL "NO" AND NOT PREBUILT_STAGE2_PATH STREQUAL "NO")
+    message(FATAL_ERROR "PREBUILT_STAGE1_PATH and PREBUILT_STAGE2_PATH is incompatible together.")
+endif()
+
+if (BUILD_STAGE1_TOOLCHAIN_ONLY OR
+    BUILD_SPIRV_TRANSLATOR_ONLY OR BUILD_VC_INTRINSICS_ONLY OR BUILD_L0_LOADER_ONLY)
+    set(XE_DEPS OFF)
+endif()
+
+if (BUILD_SPIRV_TRANSLATOR_ONLY OR BUILD_VC_INTRINSICS_ONLY OR BUILD_L0_LOADER_ONLY)
+    set(INSTALL_WITH_XE_DEPS ON)
+endif()
+
+if (APPLE)
+    # XE are not supported on macOS targets
+    set(XE_DEPS OFF)
+
+    if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
+        set (MACOS_VERSION_MIN 10.12)
+    else ()
+        set (MACOS_VERSION_MIN 11.0)
```

Comment:
Somehow stage2 always default to `CMAKE_OSX_DEPLOYMENT_TARGET` equal to `10.12`. And this is not the `10.12` set above, it's coming from somewhere else. This needs to be fixed, otherwise there's no point.

Also it doesn't need to be set for stage1, as it's use only on the host where's built.

Reply:
How do you check that?

Reply:
Ideally, when building the final ISPC binary the whole binary should be built with either `11.0` or `10.12` OSX target. it can be verified by examining the binary. But if anything goes wrong, you'll get warnings. I noticed warnings that about `10.12` not being supported for Arm, so `11.0` is used. So I investigated. Grepping for `CMAKE_OSX_DEPLOYMENT_TARGET` in build directory points to stage2 CMake that is configured with `10.12` and I've played with it to see that it's some kind of default, which is not coming from parent CMake.